examples/basic.rs

```diff
@@ -0,0 +1,128 @@
+    if let Response::Fetch(_, ref attr_vals) = *responsedata.parsed() {
+        for val in attr_vals.iter() {
+            match *val {
+                AttributeValue::Uid(u) => {
+                    eprintln!("E-mail UID: {}", u);
+                }
+                AttributeValue::Rfc822(Some(src)) => {
+                    eprintln!("E-mail body length: {}", src.to_vec().len());
+                }
+                _ => (),
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum SyncError {
+    Connect { cause: io::Error },
+    Login { cause: io::Error },
+    Select { cause: io::Error },
+    UidFetch { cause: io::Error },
+}
+
+impl Error for SyncError {
+    fn description(&self) -> &'static str {
+        ""
+    }
+
+    fn cause(&self) -> Option<&Error> {
+        match *self {
+            SyncError::Connect { ref cause }
+            | SyncError::Login { ref cause }
+            | SyncError::Select { ref cause }
+            | SyncError::UidFetch { ref cause } => Some(cause),
+        }
+    }
+}
+
+impl Display for SyncError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match *self {
+            SyncError::Connect { ref cause } => write!(f, "Connect failed: {}. ", cause),
+            SyncError::Login { ref cause } => write!(f, "Login failed: {}. ", cause),
+            SyncError::Select { ref cause } => write!(f, "Mailbox selection failed: {}. ", cause),
+            SyncError::UidFetch { ref cause } => write!(f, "Fetching e-mails failed: {}. ", cause),
+        }
+    }
+}
+
+fn main() {
+    // Provide server address, login, password and mailbox name on standard input, each on a line
+    // and 4 lines in total.
+    let (mut server, mut login, mut password, mut mailbox) =
+        (String::new(), String::new(), String::new(), String::new());
+    let (server, login, password, mailbox) = {
+        io::stdin()
+            .read_line(&mut server)
+            .expect("Provide an IMAP server FQDN. ");
+        io::stdin()
+            .read_line(&mut login)
+            .expect("Provide a login. ");
+        io::stdin()
+            .read_line(&mut password)
+            .expect("Provide a password. ");
+        io::stdin()
+            .read_line(&mut mailbox)
+            .expect("Provide a mailbox. ");
+        (server.trim(), login.trim(), password.trim(), mailbox.trim())
+    };
+    eprintln!("Will connect to {}", server);
+    eprintln!("Will login as {}", login);
+    let mut core = Core::new().expect("Failed to create a core. ");
+    let handle = core.handle();
+    let future_imap = Client::connect(server, &handle)
+        .map_err(|cause| SyncError::Connect { cause })
+        .and_then(move |(client, response)| {
+            eprintln!("Connection attempt response: {:#?}", response);
+            // todo What is the correct order of the collect and map_err calls?
+            // todo Why does client.call() not return an Err when it should, continuing the chain?
```

Comment:
I'm not sure what you mean by "when it should". The return type for `Client::call()` is just `StateStream<Item = ResponseData, State = Client, Error = io::Error>`; it can yield errors when polled, but, since all of the actual action is deferred to when the `StateStream` is polled, there will be no errors on just calling `Client::call()` itself.

Reply:
How can I stop the futures chain when e.g. the login command failed?

Reply:
Have you read the documentation on what `and_then()` means? https://docs.rs/futures/0.1.16/futures/future/trait.Future.html#method.and_then